src/apiv2.ts

```diff
@@ -0,0 +1,283 @@
+import fetch, { Response, RequestInit } from "node-fetch";
+
+import { FirebaseError } from "./error";
+import * as logger from "./logger";
+import * as responseToError from "./responseToError";
+
+const CLI_VERSION = require("../package.json").version;
+
+type HttpMethod = "GET" | "PUT" | "POST" | "DELETE" | "PATCH";
+
+interface RequestOptions<T> extends VerbOptions<T> {
+  method: HttpMethod;
+  path: string;
+  json?: T;
+  responseType?: "json";
+}
+
+interface VerbOptions<T> {
+  method?: HttpMethod;
+  headers?: { [key: string]: string };
+  queryParams?: { [key: string]: string | number };
+}
+
+interface ClientHandlingOptions {
+  skipLog?: {
+    queryParams?: boolean;
+    body?: boolean;
+    resBody?: boolean;
+  };
+  resolveOnHTTPError?: boolean;
+}
+
+export type ClientRequestOptions<T> = RequestOptions<T> & ClientVerbOptions<T>;
+
+export type ClientVerbOptions<T> = VerbOptions<T> & ClientHandlingOptions;
+
+export type ClientResponse<T> = {
+  status: number;
+  response: Response;
+  body: T;
+};
+
+let accessToken = "";
+let refreshToken = "";
+
+/**
+ * Sets the refresh token.
+ * @param token refresh token.
+ */
+export function setRefreshToken(token = ""): void {
+  refreshToken = token;
+}
+
+/**
+ * Sets the access token.
+ * @param token access token.
+ */
+export function setAccessToken(token = ""): void {
+  accessToken = token;
+}
+
+export type ClientOptions = {
+  urlPrefix: string;
+  apiVersion?: string;
+  auth?: boolean;
+};
+
+export class Client {
+  constructor(private opts: ClientOptions) {
+    if (this.opts.auth === undefined) {
+      this.opts.auth = true;
+    }
+    if (this.opts.urlPrefix.endsWith("/")) {
+      this.opts.urlPrefix = this.opts.urlPrefix.substring(0, this.opts.urlPrefix.length - 1);
```

Comment:
Isn't `substr` preferred now?

Reply:
Not that I recall. They're two different functions (which can be basically substituted here): the second argument for substr is the number is the _length_ to return, the second argument to substring is the index to stop at. Using 0 for the first argument, either is technically fine.